Upcoming releases of NVIDIA HPC SDK compilers no longer support the -tp p7-64 flag

There has been a recent change to the NVIDIA HPC SDK compilers to drop support for the `-tp p7-64` flag on x86_64 platforms.  This is tied to a recent decision to drop support in the compilers for any architectures that do not support at least AVX instructions, i.e. anything older than Intel Sandy Bridge.  OpenBLAS uses this flag to produce architecture-neutral object files when compiling with the HPC SDK compilers.

The preferred flag is now `-tp px`.

Note that the pre-AVX architecture kernels will still build with the HPC SDK compilers, but they will no longer be as useful, as the interface code may have at minimum AVX instructions in it now.

Could you update the OpenBLAS build to change `-tp p7-64` to `-tp px` now?

I can send along a patch if needed, but this should be a fairly trivial change.  Thanks in advance.

> This is actually a new change for the upcoming 21.11 release - sorry, I should have made that more clear up above. I encountered it during some of our nightly testing. I think 21.9 and earlier are fine as-is.

Thank you for the clarification - indeed the 21.9 nvc still accepts `-tp p7-64` (now that I actually installed it) although it no longer advertises that argument in its help message. 
However, what do you make of this strange ICE while compiling interface/gemm.c - is something wrong with my installation of 21.9 (on Ryzen5-4600H running OpenSUSE 15.2) or should I report this through some other channel:
```
nvc -DCBLAS -c -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -tp p7-64 -DF_INTERFACE_PGI -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.18.dev\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -DASMNAME=cblas_sgemm -DASMFNAME=cblas_sgemm_ -DNAME=cblas_sgemm_ -DCNAME=cblas_sgemm -DCHAR_NAME=\"cblas_sgemm_\" -DCHAR_CNAME=\"cblas_sgemm\" -DNO_AFFINITY -I.. -I. -UDOUBLE  -UCOMPLEX gemm.c -o cblas_sgemm.o
NVC++-F-0000-Internal compiler error. get_call_sptr(): indirect call not via LDA/ACON     154  (gemm.c: 239)
NVC++/x86-64 Linux 21.9-0: compilation aborted
make[1]: *** [Makefile:1295: sgemm.o] Error 2
make[1]: *** Waiting for unfinished jobs....
NVC++-F-0000-Internal compiler error. get_call_sptr(): indirect call not via LDA/ACON     154  (gemm.c: 438)
NVC++/x86-64 Linux 21.9-0: compilation aborted
make[1]: *** [Makefile:1883: cblas_sgemm.o] Error 2
```
(both source lines have trivial assignments to variables of type int, the first one is conditional, the second is literally just `info =-1;` , error persists with `-tp px` or `-tp zen`)
Edited to add: problem persists at `-O0` and with `TARGET=GENERIC`, i.e. just `-tp px` and none of the GCC-style sse and avx flags. Also when trying a cmake build, CMAKE fails to identify a usable CMAKE_ASM_COMPILER on its own, still fails when I supply nvc, but succeeds when I tell it to use gcc. I do not remember this being a problem for earlier versions of the HPC SDK.

> > This is actually a new change for the upcoming 21.11 release - sorry, I should have made that more clear up above. I encountered it during some of our nightly testing. I think 21.9 and earlier are fine as-is.
> 
> Thank you for the clarification - indeed the 21.9 nvc still accepts `-tp p7-64` (now that I actually installed it) although it no longer advertises that argument in its help message. However, what do you make of this strange ICE while compiling interface/gemm.c - is something wrong with my installation of 21.9 (on Ryzen5-4600H running OpenSUSE 15.2) or should I report this through some other channel:
> 
> ```
> nvc -DCBLAS -c -O2 -DSMALL_MATRIX_OPT -DMAX_STACK_ALLOC=2048 -tp p7-64 -DF_INTERFACE_PGI -fPIC -DSMP_SERVER -DNO_WARMUP -DMAX_CPU_NUMBER=12 -DMAX_PARALLEL_NUMBER=1 -DBUILD_SINGLE=1 -DBUILD_DOUBLE=1 -DBUILD_COMPLEX=1 -DBUILD_COMPLEX16=1 -DVERSION=\"0.3.18.dev\" -msse3 -mssse3 -msse4.1 -mavx -mavx2 -DASMNAME=cblas_sgemm -DASMFNAME=cblas_sgemm_ -DNAME=cblas_sgemm_ -DCNAME=cblas_sgemm -DCHAR_NAME=\"cblas_sgemm_\" -DCHAR_CNAME=\"cblas_sgemm\" -DNO_AFFINITY -I.. -I. -UDOUBLE  -UCOMPLEX gemm.c -o cblas_sgemm.o
> NVC++-F-0000-Internal compiler error. get_call_sptr(): indirect call not via LDA/ACON     154  (gemm.c: 239)
> NVC++/x86-64 Linux 21.9-0: compilation aborted
> make[1]: *** [Makefile:1295: sgemm.o] Error 2
> make[1]: *** Waiting for unfinished jobs....
> NVC++-F-0000-Internal compiler error. get_call_sptr(): indirect call not via LDA/ACON     154  (gemm.c: 438)
> NVC++/x86-64 Linux 21.9-0: compilation aborted
> make[1]: *** [Makefile:1883: cblas_sgemm.o] Error 2
> ```
> 
> (both source lines have trivial assignments to variables of type int, the first one is conditional, the second is literally just `info =-1;` , error persists with `-tp px` or `-tp zen`) Edited to add: problem persists at `-O0` and with `TARGET=GENERIC`, i.e. just `-tp px` and none of the GCC-style sse and avx flags. Also when trying a cmake build, CMAKE fails to identify a usable CMAKE_ASM_COMPILER on its own, still fails when I supply nvc, but succeeds when I tell it to use gcc. I do not remember this being a problem for earlier versions of the HPC SDK.

Hi Martin - apologies for the delayed response, but I have been having some Internet connectivity problems with my ISP for the last few days.

Yes, this ICE is a known issue.  I have filed a bug report against it internally, and it should be fixed when 21.11 comes out.  Unfortunately I do not have a good workaround for this at the moment.

I have not experimented much with using CMAKE to build here, but our compilers are built on top of LLVM.  I believe nvc passes most assembly files along to the clang assembler.  We would have to dig in further here and see why it is failing here.  I have encountered a few other bugs with inline assembly in our compilers over the years, and have filed bugs internally against them.

--- a/Makefile.system
+++ b/Makefile.system
@@ -900,16 +900,15 @@ endif
 
 ifeq ($(C_COMPILER), PGI)
 PGCVERSIONGT20 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \> 20)
-PGCVERSIONGTEQ20 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \>= 20)
-PGCMINORVERSIONGE11 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -c 4-5` == 11)
+PGCVERSIONEQ20 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` == 20)
+PGCMINORVERSIONGE11 := $(shell expr `$(CC) --version|sed -n "2p" |cut -d "-" -f 1 |sed -e "s/[^0-9.]//g" |cut -c 4-5` \>= 11)
 PGCVERSIONCHECK := $(PGCVERSIONGT20)$(PGCVERSIONEQ20)$(PGCMINORVERSIONGE11)
-ifeq ($(PGCVERSIONCHECK), $(filter $(PGCVERSIONCHECK), 110 111 011))
+ifeq ($(PGCVERSIONCHECK), $(filter $(PGCVERSIONCHECK), 100 101 011))
 NEWPGI := 1
 PGCVERSIONGT21 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \> 21)
-PGCVERSIONGTEQ21 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \>= 21)
-PGCMINORVERSIONGE9 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -c 4-5` ==9)
-PGCVERSIONCHECK2 := $(PGCVERSIONGT21)$(PGCVERSIONEQ21)$(PGCMINORVERSIONGE9)
-ifeq ($(PGCVERSIONCHECK2), $(filter $(PGCVERSIONCHECK2), 110 111 011))
+PGCVERSIONEQ21 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` == 21)
+PGCVERSIONCHECK2 := $(PGCVERSIONGT21)$(PGCVERSIONEQ21)$(PGCMINORVERSIONGE11)
+ifeq ($(PGCVERSIONCHECK2), $(filter $(PGCVERSIONCHECK2), 100 101 011))
 NEWPGI2 := 1
 endif
 endif

--- a/Makefile.system
+++ b/Makefile.system
@@ -104,7 +104,7 @@ GETARCH_FLAGS += -DUSER_TARGET
 ifeq ($(TARGET), GENERIC)
 ifeq ($(DYNAMIC_ARCH), 1)
 override NO_EXPRECISION=1
-export NO_EXPRECiSION
+export NO_EXPRECISION
 endif
 endif
 endif
@@ -905,10 +905,21 @@ PGCMINORVERSIONGE11 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-
 PGCVERSIONCHECK := $(PGCVERSIONGT20)$(PGCVERSIONEQ20)$(PGCMINORVERSIONGE11)
 ifeq ($(PGCVERSIONCHECK), $(filter $(PGCVERSIONCHECK), 110 111 011))
 NEWPGI := 1
+PGCVERSIONGT21 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \> 21)
+PGCVERSIONGTEQ21 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -d "." -f 1` \>= 21)
+PGCMINORVERSIONGE9 := $(shell expr `$(CC) --version|sed -n "2p" |sed -e "s/[^0-9.]//g" |cut -c 4-5` ==9)
+PGCVERSIONCHECK2 := $(PGCVERSIONGT21)$(PGCVERSIONEQ21)$(PGCMINORVERSIONGE9)
+ifeq ($(PGCVERSIONCHECK2), $(filter $(PGCVERSIONCHECK2), 110 111 011))
+NEWPGI2 := 1
+endif
 endif
 ifdef BINARY64
 ifeq ($(ARCH), x86_64)
+ifneq ($(NEWPGI2),1)
 CCOMMON_OPT += -tp p7-64
+else
+CCOMMON_OPT += -tp px
+endif
 ifneq ($(NEWPGI),1)
 CCOMMON_OPT +=  -D__MMX__ -Mnollvm
 endif
@@ -923,7 +934,11 @@ endif
 endif
 endif
 else
+ifneq ($(NEWPGI2),1)
 CCOMMON_OPT += -tp p7
+else
+CCOMMON_OPT += -tp px
+endif
 endif
 endif
 

